drivers/adc/ad7616/ad7616.c

```diff
@@ -38,30 +39,301 @@
+ * @param samples - Number of conversion result
+ * @param found_crc - CRC recived from the device
+ * @return 0 if the computed CRC is the same as the device, 1 otherwise
+ */
+static int32_t check_crc(struct ad7616_conversion_result *results,
+			 uint32_t samples, uint16_t found_crc)
+{
+	uint8_t computed_crc = crc_channel_pairs(results, samples);
+
+	return !(computed_crc == found_crc);
+}
+
+/**
+ * @brief Enable CRC
+ * @param dev - The device structure.
+ * @return 0 on success, an error code otherwise
+ */
+static int32_t ad7616_enable_crc(struct ad7616_dev *dev)
+{
+	int32_t ret = 0;
+
+	ret = ad7616_write_mask(dev, AD7616_REG_CONFIG, 0b11, 0b11);
+	if (ret != 0)
+		return ret;
+
+	dev->crc = 1;
+
+	return 0;
+}
+
+/**
+ * @brief Perform a toggle on the CONV GPIO and wait for the BUSY GPIO to go low.
+ *        After this, the conversion result are ready for reading
+ * @param dev - The device structure.
+ * @return 0 on success, an error code otherwise
+ */
+static int32_t ad7616_toggle_conv(struct ad7616_dev *dev)
+{
+	int ret;
+	uint8_t value;
+	uint32_t timeout = TOGGLE_DELAY;
+
+	ret = no_os_gpio_direction_input(dev->gpio_busy);
+	if (ret != 0)
+		return ret;
+
+	ret = no_os_gpio_direction_output(dev->gpio_convst, NO_OS_GPIO_HIGH);
+	if (ret != 0)
+		return ret;
+
+	// Waiting for rising edge
+	do {
+		ret = no_os_gpio_get_value(dev->gpio_busy, &value);
+		if (ret != 0)
+			return ret;
+		timeout--;
+	} while (value != NO_OS_GPIO_HIGH && timeout != 0);
+
+	if (!timeout)
+		return -ETIMEDOUT;
+	else
+		timeout = TOGGLE_DELAY;
+
+	ret = no_os_gpio_direction_output(dev->gpio_convst, NO_OS_GPIO_LOW);
+	if (ret != 0)
+		return ret;
+
+	// Waiting for falling edge
+	do {
+		ret = no_os_gpio_get_value(dev->gpio_busy, &value);
+		if (ret != 0)
+			return ret;
+		timeout--;
+	} while (value != NO_OS_GPIO_LOW && timeout != 0);
+
+	if (!timeout)
+		return -ETIMEDOUT;
+
+	return 0;
+}
+
+/**
+ * @brief Reads all conversion results with optional CRC
+ * @param dev - The device structure.
+ * @param res - Array of conversion results. The caller is responsible for allocating enough space
+ * @param crc - Pointer to store the CRC received from the device
+ * @return 0 on success, an error code otherwise
+ */
+static int32_t ad7616_read_channels(struct ad7616_dev *dev,
+				    struct ad7616_conversion_result *res,
+				    uint8_t *crc)
+{
+	int32_t ret;
+	uint32_t read_nb = dev->layers_nb * 2;
+	uint16_t tmp = 0;
+
+	for (uint32_t i = 0; i < read_nb; i++, tmp = 0) {
+		ret = no_os_spi_write_and_read(dev->spi_desc, (void *)&tmp,
+					       sizeof(tmp));
+		if (ret != 0)
+			return ret;
+
+		tmp = __builtin_bswap16(tmp);
+
+		if (i % 2 == 0)
+			res[i / 2].channel_a = tmp;
+		else
+			res[i / 2].channel_b = tmp;
+	}
+
+	// Check CRC
+	if (dev->crc) {
+		ret = no_os_spi_write_and_read(dev->spi_desc, (void *)&tmp,
+					       sizeof(tmp));
+		if (ret != 0)
+			return ret;
+
+		tmp = __builtin_bswap16(tmp);
```

Comment:
We also have no_os_get_unaligned_le16 or no_os_get_unaligned_be16 as more portable alternatives than this gcc builtin.